safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,422 @@
+/// panics and separate elections from offloading itself.
+async fn wal_backup_launcher_main_loop(
+    conf: SafeKeeperConf,
+    mut wal_backup_launcher_rx: Receiver<ZTenantTimelineId>,
+) {
+    info!(
+        "wal backup launcher started, remote config {:?}",
+        conf.remote_storage
+    );
+
+    let conf_ = conf.clone();
+    REMOTE_STORAGE.get_or_init(|| {
+        let rs = conf_.remote_storage.as_ref().map(|c| {
+            GenericRemoteStorage::new(conf_.workdir, c).expect("failed to create remote storage")
+        });
+
+        Box::new(rs)
+    });
+
+    let mut tasks: HashMap<ZTenantTimelineId, WalBackupTaskHandle> = HashMap::new();
+
+    loop {
+        // channel is never expected to get closed
+        let zttid = wal_backup_launcher_rx.recv().await.unwrap();
+        let is_wal_backup_required = is_wal_backup_required(zttid);
+        if conf.remote_storage.is_none() {
+            continue; /* just drain the channel and do nothing */
+        }
+        // do we need to do anything at all?
+        if is_wal_backup_required != tasks.contains_key(&zttid) {
+            if is_wal_backup_required {
+                // need to start the task
+                info!("starting wal backup task for {}", zttid);
+
+                // TODO: decide who should offload in launcher itself by simply checking current state
+                let election_name = broker::get_campaign_name(
+                    BACKUP_ELECTION_NAME.to_string(),
+                    conf.broker_etcd_prefix.clone(),
+                    &zttid,
+                );
+                let my_candidate_name = broker::get_candiate_name(conf.my_id);
+                let election = broker::Election::new(
+                    election_name,
+                    my_candidate_name,
+                    conf.broker_endpoints.clone(),
+                );
+
+                let (shutdown_tx, shutdown_rx) = mpsc::channel(1);
+                let timeline_dir = conf.timeline_dir(&zttid);
+
+                let handle = tokio::spawn(
+                    backup_task_main(zttid, timeline_dir, shutdown_rx, election)
+                        .instrument(info_span!("WAL backup", zttid = %zttid)),
+                );
+
+                tasks.insert(
+                    zttid,
+                    WalBackupTaskHandle {
+                        shutdown_tx,
+                        handle,
+                    },
+                );
+            } else {
+                // need to stop the task
+                info!("stopping wal backup task for {}", zttid);
+
+                let wb_handle = tasks.remove(&zttid).unwrap();
+                // Tell the task to shutdown. Error means task exited earlier, that's ok.
+                let _ = wb_handle.shutdown_tx.send(()).await;
+                // Await the task itself. TODO: restart panicked tasks earlier.
+                // Hm, why I can't await on reference to handle?
```

Comment:
You can on a `&mut` one.

Reply:
Just `&mut` didn't help -- actually awaiting on `JoinHandle` consumes it, which is reasonable. I fixed another place (killing lease task) according to this.

Reply:
Just out of curiosity, can you elaborate a bit on the consuming part?
I've meant something like 
```rust
#[tokio::main]
async fn main() {
    let mut handle = tokio::spawn(async {
        tokio::time::sleep(std::time::Duration::from_secs(3)).await;
        println!("Helloooooo from taaaaaaaaaaaask");     
    });
    
    loop {
        tokio::select! {
            _ = &mut handle => {
                println!("Task finished");
                break;
            }
            
            _ = tokio::time::sleep(std::time::Duration::from_secs(1)) => {
                println!("Waiting for the task");
            }
        }
    }
}
```

that seem to work without (explicitly?) consuming anything and awaiting on the same handle multiple times:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b83a0a089e97197c7cb270fc6cf4deda

Reply:
Interesting. Why then this doesn't work?
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ffe24232424869e889525f6e05a24262

Reply:
TIL, thanks. One liner also works: `(&mut self.h).await;`.